fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/repository/ElectionRepositoryTest.java

```diff
@@ -0,0 +1,84 @@
+import com.github.dedis.popstellar.model.objects.Election;
+import com.github.dedis.popstellar.model.objects.Lao;
+import com.github.dedis.popstellar.model.objects.event.EventState;
+import com.github.dedis.popstellar.utility.error.UnknownElectionException;
+
+import org.junit.Test;
+
+import java.util.Set;
+
+import io.reactivex.observers.TestObserver;
+
+import static com.github.dedis.popstellar.testutils.Base64DataUtils.generatePublicKey;
+import static java.util.Collections.emptySet;
+import static java.util.Collections.singleton;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.junit.Assert.assertThrows;
+
+public class ElectionRepositoryTest {
+
+  private static final String LAO_ID = Lao.generateLaoId(generatePublicKey(), 100321004, "Lao");
+  private static final Election ELECTION =
+      new Election.ElectionBuilder(LAO_ID, 100321014, "Election")
+          .setElectionVersion(ElectionVersion.OPEN_BALLOT)
+          .build();
+
+  @Test
+  public void addingElectionUpdatesIds() {
+    ElectionRepository repo = new ElectionRepository();
+    TestObserver<Set<String>> ids = repo.getElectionsObservable(LAO_ID).test();
+
+    assertCurrentValueIs(ids, emptySet());
+
+    repo.updateElection(ELECTION);
+
+    assertCurrentValueIs(ids, singleton(ELECTION.getId()));
+  }
+
+  @Test
+  public void electionUpdateIsCorrectlyDispatched() throws UnknownElectionException {
+    // Create repository with an election inside it
+    ElectionRepository repo = new ElectionRepository();
+    repo.updateElection(ELECTION);
+    TestObserver<Election> electionObserver =
+        repo.getElectionObservable(LAO_ID, ELECTION.getId()).test();
+
+    assertCurrentValueIs(electionObserver, ELECTION);
+    assertThat(repo.getElection(LAO_ID, ELECTION.getId()), is(ELECTION));
+
+    // Update Election
+    Election updated = ELECTION.builder().setState(EventState.CLOSED).build();
+    repo.updateElection(updated);
+
+    // Assert that the update was correctly dispatched
+    assertCurrentValueIs(electionObserver, updated);
+    assertThat(repo.getElection(LAO_ID, ELECTION.getId()), is(updated));
+  }
+
+  @Test
+  public void electionByChannelHasSameEffectAsGetElection() throws UnknownElectionException {
+    ElectionRepository repo = new ElectionRepository();
+    repo.updateElection(ELECTION);
+
+    assertThat(
+        repo.getElection(LAO_ID, ELECTION.getId()),
+        is(repo.getElectionByChannel(ELECTION.getChannel())));
+  }
+
+  @Test
+  public void retrievingAnInvalidElectionThrowsAnException() {
+    ElectionRepository repo = new ElectionRepository();
+
+    assertThrows(UnknownElectionException.class, () -> repo.getElection(LAO_ID, ELECTION.getId()));
+    assertThrows(
+        UnknownElectionException.class, () -> repo.getElectionObservable(LAO_ID, ELECTION.getId()));
+  }
+
+  private <T> void assertCurrentValueIs(TestObserver<T> observer, T expected) {
```

Comment:
In the digital cash decoupling PR, I moved that function that you defined in the social media decoupling PR to a `ObservableUtils` class. So there is no need to redefine it again 🙂

Reply:
Oh right, I'll do that